[FSDP] Parameter shape verification and synchronization

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75977
* __->__ #75976
* #75970
* #75945

Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

rohan-varma committed:
[FSDP] Parameter shape verification and synchronization

Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -461,6 +465,31 @@ def __init__(
+        was_cpu = (prev_device == torch.device("cpu"))
+        if was_cpu:
+            warnings.warn(
+                f"Module is input on CPU, we are moving to to {self.compute_device} to perform parameter verification and will move it back after. In the future please specify device_ids which will explicitly tell FSDP to place the module on a particular device."
```
nit: `to to` -> `to`

`device_ids` -> `device_id`? (We want to enforce SPSD for FSDP?)

mrshenli replied:
Is it OK to mention `device_id` at this point, given that this arg is not available yet?

rohan-varma replied:
Was planning to base the next PR under this one and land this after that one lands and we agree on CPU-side semantics. Currently PR is mostly for discussion

rohan-varma replied:
Will remove it for now, and update warning when device_ids PR lands.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -461,6 +465,31 @@ def __init__(
+        was_cpu = (prev_device == torch.device("cpu"))
+        if was_cpu:
+            warnings.warn(
+                f"Module is input on CPU, we are moving to to {self.compute_device} to perform parameter verification and will move it back after. In the future please specify device_ids which will explicitly tell FSDP to place the module on a particular device."
```
Is `self.compute_device` guaranteed to be a CUDA device (and not CPU)?

mrshenli replied:
line too long, shall we break it into shorter ones?

rohan-varma replied:
@awgu Yep it is

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -461,6 +465,31 @@ def __init__(
+            module = module.to(self.compute_device)
+        _verify_param_shape_across_processes(
+            process_group=self.process_group,
+            tensors=params,  # TODO: FSDP + DDP to verify buffers as well
```
My understanding is that shared parameters will be verified and sync-ed multiple times, but there is not an easy way around this. Is this correct?

rohan-varma replied:
Yeah, hopefully there are not too many shared params where it becomes a perf issue.

awgu reviewed (approved):
Stamping to unblock since the logic all makes sense to me. Left some minor comments.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -461,6 +465,31 @@ def __init__(
+            rank=0,
+            params_and_buffers_to_ignore=ignored_params,
+        )
+        if was_cpu:
+            module = module.to(prev_device)
```
Could we move this to be after "self._shard_parameters()", so that we can flatten and shard parameters on GPU, this will greatly improve model init time when users pass a CPU model without device_id specified?

rohan-varma replied:
This is taken care of if users pass in `device_id`s already. Also this PR and `was_cpu` logic is mostly intended for discussion, as this does not seem like a great way to support CPU use cases?

zhaojuanmao replied:
that sounds good we only support GPU sharding when users pass a device id. although we can support GPU sharding and move sharded module from GPU to CPU even users do not pass a device id? but probably it is not a good user pattern...

rohan-varma replied:
Sure, per our discussion we can offer basic support for this and log a warning saying device_ids should be used. Will add warning in the next diff for device_Ids.

For this PR, if CPU, we'll move to compute_device (in the future compute_device == device_id should be an invariant), and run sharding, flattening as well on GPU before moving the module back to CPU.

rohan-varma replied:
We also call `del module` after sharding and before flattening, I think given this, if original model was on CPU, we can simply move it back similar to how we offload the param to CPU in ctor.

rohan-varma committed:
Update on "[FSDP] Parameter shape verification and synchronization"


Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

[ghstack-poisoned]

zhaojuanmao commented:
sorry, we need to benchmark this PR before landing

rohan-varma committed:
Update on "[FSDP] Parameter shape verification and synchronization"


Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Parameter shape verification and synchronization"


Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Parameter shape verification and synchronization"


Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Parameter shape verification and synchronization"


Use generalized APIs to perform param shape verification and module
synchronization.

Differential Revision: [D35712026](https://our.internmc.facebook.com/intern/diff/D35712026/)

[ghstack-poisoned]

rohan-varma commented:
Close in favor of https://github.com/pytorch/pytorch/pull/76442/files
